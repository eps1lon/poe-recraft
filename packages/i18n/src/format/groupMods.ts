import formatStats, { Options as FormatStatsOptions, Stat } from './stats';

// arg types
export { Stat } from './stats';

export type Options = {
  resolveWordConflict: (words: string[]) => string;
} & FormatStatsOptions;

/**
 * tries to find a string that describes the given mods
 *
 * given a list of mods where a mod consists of at least one stat
 * get a translation t of that mod and consider it a row
 * split t into words and consider every word as a column
 * collapse the table into a single row
 * columns with different words get resolved by a given strategy
 *
 *
 *
 * Adds # to Fire Gems
 * Adds # to Cold Gems
 * ----
 * Adds # to * Gems
 *
 * @param mods
 * @param options
 */
export default function groupMod(
  mods: Stat[][],
  options: Partial<Options> = {},
): string {
  // default options
  const { resolveWordConflict = () => '*', ...format_stats_options } = options;

  // reduce mods to array
  const translations: string[][] = [];
  for (const mod of mods) {
    translations.push(groupStats(mod, format_stats_options).split(' '));
  }

  return collapseTable(translations, resolveWordConflict)
    .join(' ')
    .replace(/\*( \*)*/, '*');
}

function groupStats(
  stats: Stat[],
  options: Partial<FormatStatsOptions> = {},
): string {
  const lines = formatStats(stats, {
    ...options,
    getFormatters: (_t, _s, n) => {
      return Array.from({ length: n }, (_, i) => ({
        arg: i + 1,
        id: 'placeholder',
      }));
    },
  });

  // collapes value ranges into single placeholder
  return lines.map(line => line.replace(/\(# - #\)/g, '#')).join(' / ');
}

function collapseTable<T>(
  table: T[][],
  resolveColumnConflict: (items: T[]) => T,
): T[] {
  if (table.length < 1) {
    return [];
  }

  const column_count = table[0].length;

  // rows to columns
  const columns: Array<Set<T>> = new Array(column_count);
  for (let j = 0; j < column_count; ++j) {
    columns[j] = table.reduce((column, row) => {
      return column.add(row[j]);
    }, new Set<T>());
  }

  return columns.map(column => {
    if (column.size > 1) {
      return resolveColumnConflict(Array.from(column));
    } else {
      return column.values().next().value;
    }
  });
}
