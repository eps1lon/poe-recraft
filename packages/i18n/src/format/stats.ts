import translate, { NO_DESCRIPTION, Stat } from '../translate';
import { ICUMessageSyntax } from '../types/intl';
import {
  Description,
  Descriptions,
  StatLocaleData,
  StatLocaleDatas,
  Translation,
} from '../types/StatDescription';
import { isZero } from '../types/StatValue';

// arg types
export { Stat } from '../translate';
export const DEFAULT_RANGE_MESSAGE = '({min}â€“{max})';
export type Options = {
  datas: StatLocaleDatas;
  fallback: Fallback | FallbackCallback;
  start_file: string;
  getFormatters: (
    t: Translation,
    stat: Stat,
    n: number,
  ) => Translation['formatters'];
  /**
   * if a stat value is rollable (i.e. has a min and max value)
   * default: {DEFAULT_RANGE_MESSAGE}
   */
  range_message: ICUMessageSyntax;
};
// return type
export type TranslatedStats = string[];

export type FallbackCallback = (id: string, stat: Stat) => string | null;

export class NoDescriptionFound extends Error {
  constructor(stats: Stat[]) {
    super('no descriptions found for ' + stats.map(({ id }) => id).join(','));
  }
}
export enum Fallback {
  throw, // throw if no stat was found
  id,
  skip,
}

const initial_options: Options = {
  datas: {},
  fallback: Fallback.throw,
  start_file: 'stat_descriptions',
  getFormatters: t => t.formatters,
  range_message: DEFAULT_RANGE_MESSAGE,
};

const formatStats = (
  stats: Stat[],
  options: Partial<Options> = {},
): TranslatedStats => {
  const {
    datas,
    fallback,
    start_file,
    getFormatters,
    range_message,
  } = Object.assign({}, initial_options, options);

  // translated lines
  const lines: string[] = [];
  // array of stat_ids for which hash lookup failed
  const untranslated: Map<string, Stat> = new Map(
    stats.map((stat: Stat) => [stat.id, stat] as [string, Stat]),
  );

  let description_file: StatLocaleData | undefined = datas[start_file];

  while (description_file !== undefined) {
    const data: Descriptions = description_file.data;

    for (const descriptionFinder of createDescriptionFindStrategies(data)) {
      lines.push(
        ...formatWithFinder(untranslated, descriptionFinder, {
          getFormatters,
          range_message,
        }),
      );
    }

    description_file = description_file.meta.include
      ? datas[description_file.meta.include]
      : undefined;
  }

  lines.push(...formatWithFallback(untranslated, fallback));

  return lines;
};

export default formatStats;

/**
 * creates an array of methods that can be used to find a description for a
 * given stat.
 *
 * return value is to be interpreted as a priority queue
 * @param descriptions
 */
export function createDescriptionFindStrategies(
  descriptions: Descriptions,
): Array<(stat: Stat) => Description | undefined> {
  return [
    ({ id }) => descriptions[id],
    ({ id }) =>
      Object.values(descriptions).find(({ stats }) => stats.includes(id)),
  ];
}

// stats will get mutated
interface FormatWithFinderOptions {
  getFormatters: (
    t: Translation,
    stat: Stat,
    n: number,
  ) => Translation['formatters'];
  range_message: ICUMessageSyntax;
}
function formatWithFinder(
  stats: Map<string, Stat>,
  find: (stat: Stat) => Description | undefined,
  options: Partial<FormatWithFinderOptions> = {},
): string[] {
  const {
    getFormatters = (t: Translation) => t.formatters,
    range_message = DEFAULT_RANGE_MESSAGE,
  } = options;
  const lines: string[] = [];
  const translated: Set<string> = new Set();

  for (const [stat_id, stat] of Array.from(stats.entries())) {
    if (translated.has(stat_id)) {
      continue;
    }

    const description = find(stat);

    if (description !== undefined) {
      const translation = translate(
        description,
        stats,
        (t: Translation, n) => getFormatters(t, stat, n),
        range_message,
      );

      if (translation === undefined) {
        const requiredStatsAreZero = requiredStats(description, stats).every(
          ({ value }) => isZero(value),
        );

        if (!requiredStatsAreZero) {
          throw new Error(`matching translation not found for '${stat.id}'`);
        }
      } else {
        // mark as translated
        description.stats.forEach(translated_id => {
          stats.delete(translated_id);
          translated.add(translated_id);
        });

        if (translation === NO_DESCRIPTION) {
          lines.push(`${stat_id} (hidden)`);
        } else {
          lines.push(translation);
        }
      }
    }
  }

  return lines;
}

function requiredStats(
  description: Description,
  provided: Map<string, Stat>,
): Stat[] {
  // intersect the required stat_ids from the desc with the provided
  return description.stats
    .map(stat_id => {
      const stat = provided.get(stat_id);

      // default the value to 0
      if (stat === undefined) {
        return {
          id: stat_id,
          value: 0,
        };
      } else {
        return stat;
      }
    })
    .filter((stat: Stat | null): stat is Stat => stat !== null);
}

function formatWithFallback(
  stats: Map<string, Stat>,
  fallback: Fallback | FallbackCallback,
): string[] {
  const non_zero_stats = Array.from(stats.entries()).filter(
    ([, stat]) => !isZero(stat.value),
  );
  if (non_zero_stats.length === 0) {
    return [];
  }

  if (fallback === Fallback.throw) {
    if (stats.size > 0) {
      throw new NoDescriptionFound(non_zero_stats.map(([, stat]) => stat));
    } else {
      return [];
    }
  } else if (fallback === Fallback.id) {
    return non_zero_stats.map(([key]) => key);
  } else if (fallback === Fallback.skip) {
    return [];
  } else if (typeof fallback === 'function') {
    return non_zero_stats
      .map(([id, stat]) => fallback(id, stat))
      .filter((line): line is string => typeof line === 'string');
  } else {
    // should ts recognize that this is unreachable code? enums can prob
    // be extended at runtime an therfore somebody could mess with them
    throw new Error(`unrecognized fallback type '${fallback}'`);
  }
}
