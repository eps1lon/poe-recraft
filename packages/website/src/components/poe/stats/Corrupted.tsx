import React, { SFC } from 'react';

const Corrupted: SFC = () => {
  return <span className="text-corrupted">Corrupted</span>;
};

export default Corrupted;
