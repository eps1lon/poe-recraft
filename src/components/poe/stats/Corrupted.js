import React from 'react';

export type Props = {};

const Corrupted = (props: Props) => {
  return <span class="text-corrupted">Corrupted</span>;
};

export default Corrupted;
