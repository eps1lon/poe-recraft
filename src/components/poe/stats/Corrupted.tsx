import React, { SFC } from 'react';

export type Props = {};

const Corrupted: SFC<Props> = props => {
  return <span className="text-corrupted">Corrupted</span>;
};

export default Corrupted;
