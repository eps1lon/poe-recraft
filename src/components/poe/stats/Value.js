// @flow
import React from 'react';

// TODO acceppt only number and format accordingly
export type Props = {
  value: string
};

const Value = ({ value }: Props) => {
  return <span className="text-value">{value}</span>;
};

export default Value;
