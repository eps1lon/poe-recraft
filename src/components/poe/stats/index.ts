export { default as Corrupted } from './Corrupted';
export { default as Value } from './Value';
