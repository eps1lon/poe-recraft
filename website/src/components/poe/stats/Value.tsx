import React, { SFC } from 'react';

// TODO acceppt only number and format accordingly
export interface Props {
  value: string;
}

const Value: SFC<Props> = ({ value }) => {
  return <span className="text-value">{value}</span>;
};

export default Value;
