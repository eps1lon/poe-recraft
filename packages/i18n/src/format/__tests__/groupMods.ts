import { createMods, Mod } from 'poe-mods';

import datas from '../../__fixtures__/english';
import groupMods, { Options, Stat } from '../groupMods';
import { Fallback } from '../stats';

it('merges multiple mods (mod = one ore more stats) into a single line', () => {
  expect(
    groupMods([[{ id: 'weapon_physical_damage_+%', value: 25 }]], { datas })
  ).toEqual('#% increased Physical Damage with Weapons');

  expect(
    groupMods(
      [
        [
          { id: 'attack_minimum_added_physical_damage', value: 1 },
          { id: 'attack_maximum_added_physical_damage', value: 5 }
        ],

        [
          { id: 'attack_minimum_added_physical_damage', value: 10 },
          { id: 'attack_maximum_added_physical_damage', value: 20 }
        ]
      ],
      { datas }
    )
  ).toEqual('Adds # to # Physical Damage to Attacks');
});

it('merges multiple stats within a mod by "/"', () => {
  expect(
    groupMods(
      [
        [
          { id: 'local_energy_shield_+%', value: 5 },
          { id: 'base_stun_recovery_+%', value: 10 }
        ]
      ],
      { datas }
    )
  ).toEqual(
    '#% increased Energy Shield / #% increased Stun and Block Recovery'
  );
});

it('its resolve conflict strategy can be configured', () => {
  expect(
    groupMods(
      [
        [{ id: 'local_socketed_fire_gem_level_+', value: 1 }],

        [{ id: 'local_socketed_cold_gem_level_+', value: 10 }]
      ],
      { datas }
    )
  ).toEqual('+# to Level of Socketed * Gems');
});

it('returns an emtpy string if no mods were given', () => {
  expect(groupMods([])).toEqual('');
});

it('merges multiple mods (mod = one ore more stats) into a single line', () => {
  expect(
    groupMods([[{ id: 'weapon_physical_damage_+%', value: 25 }]], { datas })
  ).toEqual('#% increased Physical Damage with Weapons');

  expect(
    groupMods(
      [
        [
          { id: 'attack_minimum_added_physical_damage', value: [1, 5] },
          { id: 'attack_maximum_added_physical_damage', value: [10, 15] }
        ],

        [
          { id: 'attack_minimum_added_physical_damage', value: [6, 10] },
          { id: 'attack_maximum_added_physical_damage', value: [16, 20] }
        ]
      ],
      { datas }
    )
  ).toEqual('Adds # to # Physical Damage to Attacks');
});

describe('usage with poe-mods', () => {
  const modStats = (mod: Mod['props']) =>
    mod.stats.map((stat, i) => {
      const key_min = `stat${i + 1}_min` as keyof typeof mod;
      const key_max = `stat${i + 1}_min` as keyof typeof mod;

      return {
        id: stat.id,
        value: [+mod[key_min], +mod[key_max]] as [number, number]
      };
    });
  const groupByCorrectGroup = (mods: Array<Mod['props']>) => {
    return mods.reduce((groups, mod) => {
      if (!groups.has(mod.correct_group)) {
        groups.set(mod.correct_group, []);
      }
      groups.get(mod.correct_group)!.push(modStats(mod));
      return groups;
    }, new Map<string, Stat[][]>());
  };

  const locale_data = {
    stat_descriptions: require('../../../locale-data/en/stat_descriptions.json')
  };

  const options: Partial<Options> = {
    datas: locale_data,
    fallback: Fallback.skip
  };

  const prefixes = createMods(require('poe-mods/data/mods/prefixes.json'));
  // skip monster domain
  const prefix_groups = groupByCorrectGroup(
    prefixes.all().filter(mod => mod.domain === 1)
  );
  for (const [correct_group, mods] of prefix_groups.entries()) {
    it(`matches the snapshot of '${correct_group}'`, () => {
      expect(groupMods(mods, options)).toMatchSnapshot();
    });
  }

  it('groups matching words', () => {
    const group = prefixes
      .all()
      .filter(mod => mod.correct_group === 'AreaOfEffect');
    const t = groupMods(group.map(modStats), options);
    expect(t).toEqual('#% increased Area of Effect *');
  });
});
