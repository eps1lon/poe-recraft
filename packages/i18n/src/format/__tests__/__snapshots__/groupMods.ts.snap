// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usage with poe-mods matches the snapshot of 'AddedColdDamagePerDexterity' 1`] = `"Adds # to # Cold Damage to Attacks with this Weapon per 10 Dexterity"`;

exports[`usage with poe-mods matches the snapshot of 'AddedColdDamagePerFrenzyCharge' 1`] = `"# to # Cold Damage per Frenzy Charge"`;

exports[`usage with poe-mods matches the snapshot of 'AddedColdDamageToSpellsAndAttacks' 1`] = `"Adds # to # Cold Damage to Spells and Attacks"`;

exports[`usage with poe-mods matches the snapshot of 'AddedFireDamagePerStrength' 1`] = `"Adds # to # Fire Damage to Attacks with this Weapon per 10 Strength"`;

exports[`usage with poe-mods matches the snapshot of 'AddedFireDamageSpellsAndAttacks' 1`] = `"Adds # to # Fire Damage to Spells and Attacks"`;

exports[`usage with poe-mods matches the snapshot of 'AddedLightningDamagePerIntelligence' 1`] = `"Adds # to # Lightning Damage to Attacks with this Weapon per 10 Intelligence"`;

exports[`usage with poe-mods matches the snapshot of 'AddedLightningDamageSpellsAndAttacks' 1`] = `"Adds # to # Lightning Damage to Spells and Attacks"`;

exports[`usage with poe-mods matches the snapshot of 'AdditionalArrows' 1`] = `"Bow Attacks fire an additional Arrow"`;

exports[`usage with poe-mods matches the snapshot of 'AdditionalCurseOnEnemies' 1`] = `"You can apply an additional Curse"`;

exports[`usage with poe-mods matches the snapshot of 'AilmentChanceAddedDamage' 1`] = `"#% chance to * / Adds # to # * Damage against * *"`;

exports[`usage with poe-mods matches the snapshot of 'AllDamage' 1`] = `"#% increased Damage"`;

exports[`usage with poe-mods matches the snapshot of 'AlwaysHits' 1`] = `"Hits can't be Evaded"`;

exports[`usage with poe-mods matches the snapshot of 'AreaDamage' 1`] = `"#% increased Area Damage / Socketed Gems are Supported by Level # Concentrated Effect"`;

exports[`usage with poe-mods matches the snapshot of 'AreaOfEffect' 1`] = `"#% increased Area of Effect *"`;

exports[`usage with poe-mods matches the snapshot of 'AttackDamage' 1`] = `"#% increased Attack Damage"`;

exports[`usage with poe-mods matches the snapshot of 'AttackerTakesDamageNoRange' 1`] = `"Reflects # Physical Damage to Melee Attackers"`;

exports[`usage with poe-mods matches the snapshot of 'BaseLocalDefences' 1`] = `"+# to *"`;

exports[`usage with poe-mods matches the snapshot of 'BaseLocalDefencesAndLife' 1`] = `"+# to *"`;

exports[`usage with poe-mods matches the snapshot of 'BeltFlaskRecoveryRate' 1`] = `"#% increased Flask * Recovery rate"`;

exports[`usage with poe-mods matches the snapshot of 'CannotBeFrozen' 1`] = `"Cannot be Frozen"`;

exports[`usage with poe-mods matches the snapshot of 'CausesBleeding' 1`] = `"* chance to cause Bleeding on Hit"`;

exports[`usage with poe-mods matches the snapshot of 'ChaosDamageLifeLeech' 1`] = `"#% of Chaos Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'ChaosDamageTakenPercentage' 1`] = `"#% reduced Chaos Damage taken"`;

exports[`usage with poe-mods matches the snapshot of 'ColdDamage' 1`] = `"Adds # to # Cold Damage *"`;

exports[`usage with poe-mods matches the snapshot of 'ColdDamageLifeLeech' 1`] = `"#% of Cold Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'ColdDamageTaken' 1`] = `"#% reduced Cold Damage taken"`;

exports[`usage with poe-mods matches the snapshot of 'ColdResistancePenetration' 1`] = `"Damage Penetrates #% Cold Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'DamageDuringFlaskEffect' 1`] = `"#% increased Damage during any Flask Effect"`;

exports[`usage with poe-mods matches the snapshot of 'DamageOverTimeMultiplier' 1`] = `"+#% to *"`;

exports[`usage with poe-mods matches the snapshot of 'DamagePer15Attributes' 1`] = `"#% increased Damage per 15 *"`;

exports[`usage with poe-mods matches the snapshot of 'DamagePerBlockChance' 1`] = `"#% increased Damage per 1% Chance to Block Attack Damage"`;

exports[`usage with poe-mods matches the snapshot of 'DamageRemovedFromManaBeforeLife' 1`] = `"#% of Damage is taken from Mana before Life"`;

exports[`usage with poe-mods matches the snapshot of 'DamageToEnemiesPerCurse' 1`] = `"#% increased Damage with Hits and Ailments per Curse on Enemy"`;

exports[`usage with poe-mods matches the snapshot of 'DefencesPercent' 1`] = `"#% increased *"`;

exports[`usage with poe-mods matches the snapshot of 'DefencesPercentAndStunRecovery' 1`] = `"#% increased *"`;

exports[`usage with poe-mods matches the snapshot of 'DisplaySocketedGemGetsFortify' 1`] = `"Socketed Gems are Supported by Level # Fortify"`;

exports[`usage with poe-mods matches the snapshot of 'ElementalDamagePercent' 1`] = `"#% increased Elemental Damage"`;

exports[`usage with poe-mods matches the snapshot of 'ElementalDamagePercentAddedAsChaos' 1`] = `"Gain #% of Elemental Damage as Extra Chaos Damage"`;

exports[`usage with poe-mods matches the snapshot of 'ElementalPenetration' 1`] = `"Damage Penetrates #% Elemental Resistances"`;

exports[`usage with poe-mods matches the snapshot of 'EnduranceChargeOnKillChance' 1`] = `"#% chance to gain an Endurance Charge on Kill"`;

exports[`usage with poe-mods matches the snapshot of 'EnergyShieldPercent' 1`] = `"#% increased maximum Energy Shield"`;

exports[`usage with poe-mods matches the snapshot of 'EvasionRatingPercent' 1`] = `"#% increased Evasion Rating"`;

exports[`usage with poe-mods matches the snapshot of 'ExperienceIncrease' 1`] = `"#% increased Experience gain"`;

exports[`usage with poe-mods matches the snapshot of 'FireDamage' 1`] = `"Adds # to # Fire Damage *"`;

exports[`usage with poe-mods matches the snapshot of 'FireDamageAsPortionOfDamage' 1`] = `"Gain #% of Physical Damage as Extra Fire Damage"`;

exports[`usage with poe-mods matches the snapshot of 'FireDamageLifeLeech' 1`] = `"#% of Fire Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'FireDamageTaken' 1`] = `"#% reduced Fire Damage taken"`;

exports[`usage with poe-mods matches the snapshot of 'FireResistancePenetration' 1`] = `"Damage Penetrates #% Fire Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'FishingLineStrength' 1`] = `"#% increased Fishing Line Strength"`;

exports[`usage with poe-mods matches the snapshot of 'FishingLureType' 1`] = `"Rhoa Feather Lure"`;

exports[`usage with poe-mods matches the snapshot of 'FishingPoolConsumption' 1`] = `"#% reduced Fishing Pool Consumption"`;

exports[`usage with poe-mods matches the snapshot of 'FlaskChargesUsed' 1`] = `"#% chance for your Flasks to not consume Charges"`;

exports[`usage with poe-mods matches the snapshot of 'FlaskEffect' 1`] = `"#% increased effect of Flasks on you"`;

exports[`usage with poe-mods matches the snapshot of 'FortifyOnMeleeStun' 1`] = `"#% chance to gain Fortify when you Stun an Enemy with Melee Damage"`;

exports[`usage with poe-mods matches the snapshot of 'FrenzyChargeOnKillChance' 1`] = `"#% chance to gain a Frenzy Charge on Kill"`;

exports[`usage with poe-mods matches the snapshot of 'GainArmourIfBlockedRecently' 1`] = `"+# Armour if you've Blocked Recently"`;

exports[`usage with poe-mods matches the snapshot of 'GrantedSkill' 1`] = `"Grants Level # * Cry Skill"`;

exports[`usage with poe-mods matches the snapshot of 'GroundEffectEffectiveness' 1`] = `"Unaffected by * Ground"`;

exports[`usage with poe-mods matches the snapshot of 'IncreaseFlatManaCost' 1`] = `"+# to Total Mana Cost of Skills"`;

exports[`usage with poe-mods matches the snapshot of 'IncreaseSkillGemLevel' 1`] = `"+# to Level of Socketed Active Skill Gems"`;

exports[`usage with poe-mods matches the snapshot of 'IncreaseSocketedGemLevel' 1`] = `"+# to Level of Socketed Gems"`;

exports[`usage with poe-mods matches the snapshot of 'IncreaseSpecificSocketedGemLevel' 1`] = `"+# to Level of Socketed * Gems"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedEnergyShield' 1`] = `"+# to maximum Energy Shield"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedEvasionRating' 1`] = `"+# to Evasion Rating"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedLife' 1`] = `"+# to maximum Life"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedMana' 1`] = `"+# to maximum Mana"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedPhysicalDamageReductionRating' 1`] = `"+# to Armour"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedPhysicalDamageReductionRatingPercent' 1`] = `"#% increased Armour"`;

exports[`usage with poe-mods matches the snapshot of 'IncreasedWeaponElementalDamagePercent' 1`] = `"#% increased Elemental Damage with Attack Skills"`;

exports[`usage with poe-mods matches the snapshot of 'ItemFoundRarityIncreasePrefix' 1`] = `"#% increased Rarity of Items found"`;

exports[`usage with poe-mods matches the snapshot of 'LifeGainPerTarget' 1`] = `"+# Life gained for each Enemy hit by your Attacks"`;

exports[`usage with poe-mods matches the snapshot of 'LifeGainedOnSpellHit' 1`] = `"+# Life gained for each Enemy hit by your Spells"`;

exports[`usage with poe-mods matches the snapshot of 'LifeLeech' 1`] = `"#% of Physical Attack Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'LifeLeechSpeed' 1`] = `"#% increased Life Leeched per second"`;

exports[`usage with poe-mods matches the snapshot of 'LightningDamage' 1`] = `"Adds # to # Lightning Damage *"`;

exports[`usage with poe-mods matches the snapshot of 'LightningDamageLifeLeech' 1`] = `"#% of Lightning Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'LightningDamageTaken' 1`] = `"#% reduced Lightning Damage taken"`;

exports[`usage with poe-mods matches the snapshot of 'LightningResistancePenetration' 1`] = `"Damage Penetrates #% Lightning Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'LocalChaosDamage' 1`] = `"Adds # to # Chaos Damage"`;

exports[`usage with poe-mods matches the snapshot of 'LocalChaosDamageHybrid' 1`] = `"Adds # to # Chaos Damage / #% increased Chaos Damage"`;

exports[`usage with poe-mods matches the snapshot of 'LocalColdDamageHybrid' 1`] = `"Adds # to # Cold Damage / #% increased Cold Damage"`;

exports[`usage with poe-mods matches the snapshot of 'LocalFireDamageHybrid' 1`] = `"Adds # to # Fire Damage / #% increased Fire Damage"`;

exports[`usage with poe-mods matches the snapshot of 'LocalIncreasedPhysicalDamagePercentAndAccuracyRating' 1`] = `"#% increased Physical Damage / +# to Accuracy Rating"`;

exports[`usage with poe-mods matches the snapshot of 'LocalLightningDamageHybrid' 1`] = `"Adds # to # Lightning Damage / #% increased Lightning Damage"`;

exports[`usage with poe-mods matches the snapshot of 'LocalPhysicalDamagePercent' 1`] = `"#% increased Physical Damage"`;

exports[`usage with poe-mods matches the snapshot of 'ManaGainPerTarget' 1`] = `"+# Mana gained for each Enemy hit by your Attacks"`;

exports[`usage with poe-mods matches the snapshot of 'ManaGainedOnSpellHit' 1`] = `"+# Mana gained for each Enemy Hit by your Spells"`;

exports[`usage with poe-mods matches the snapshot of 'ManaLeech' 1`] = `"#% of Physical Attack Damage Leeched as Mana"`;

exports[`usage with poe-mods matches the snapshot of 'MapInvasionMonsterPacks' 1`] = `"Area is inhabited by an additional Invasion Boss / map_item_drop_quantity_+% (hidden)"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumChaosResistance' 1`] = `"+#% to maximum Chaos Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumColdResist' 1`] = `"+#% to maximum Cold Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumFireResist' 1`] = `"+#% to maximum Fire Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumLeechRate' 1`] = `"+#% of maximum Life per second to maximum Life Leech rate"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumLifeIncreasePercent' 1`] = `"#% increased maximum Life"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumLightningResistance' 1`] = `"+#% to maximum Lightning Resistance"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumManaIncreasePercent' 1`] = `"#% increased maximum Mana"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumMinionCount' 1`] = `"+# to maximum number of Zombies / +# to maximum number of Skeletons / +# to maximum number of Spectres"`;

exports[`usage with poe-mods matches the snapshot of 'MaximumResistances' 1`] = `"+#% to all maximum Resistances"`;

exports[`usage with poe-mods matches the snapshot of 'MinionDamage' 1`] = `"Minions deal #% increased Damage / Socketed Gems are Supported by Level # Minion Damage"`;

exports[`usage with poe-mods matches the snapshot of 'MinionLife' 1`] = `"Minions have #% increased maximum Life / Socketed Gems are Supported by Level # Minion Life"`;

exports[`usage with poe-mods matches the snapshot of 'MovementVelocity' 1`] = `"#% increased Movement Speed"`;

exports[`usage with poe-mods matches the snapshot of 'NonChaosAddedAsChaos' 1`] = `"Gain #% of Non-Chaos Damage as extra Chaos Damage"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalAddedAsCold' 1`] = `"Gain #% of Physical Damage as Extra Cold Damage"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalAddedAsFire' 1`] = `"Gain #% of Physical Damage as Extra Fire Damage"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalAddedAsLightning' 1`] = `"Gain #% of Physical Damage as Extra Lightning Damage"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalDamage' 1`] = `"Adds # to # Physical Damage *"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalDamageLifeLeech' 1`] = `"#% of Physical Damage Leeched as Life"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalDamagePercent' 1`] = `"#% increased Global Physical Damage"`;

exports[`usage with poe-mods matches the snapshot of 'PhysicalDamageTakenAsElement' 1`] = `"#% of Physical Damage from Hits taken as * Damage"`;

exports[`usage with poe-mods matches the snapshot of 'Pierce' 1`] = `"Projectiles Pierce * additional *"`;

exports[`usage with poe-mods matches the snapshot of 'PoisonDamage' 1`] = `"#% increased Damage with Poison"`;

exports[`usage with poe-mods matches the snapshot of 'PowerChargeOnKillChance' 1`] = `"#% chance to gain a Power Charge on Kill"`;

exports[`usage with poe-mods matches the snapshot of 'ProjectileDamage' 1`] = `"#% increased Projectile Damage / Socketed Gems are Supported by Level # Slower Projectiles"`;

exports[`usage with poe-mods matches the snapshot of 'ProjectileSpeed' 1`] = `"#% increased Projectile Speed / Socketed Gems are supported by Level # Faster Projectiles"`;

exports[`usage with poe-mods matches the snapshot of 'RarityDuringFlaskEffect' 1`] = `"#% increased Rarity of Items found during any Flask Effect"`;

exports[`usage with poe-mods matches the snapshot of 'ReducedCriticalStrikeDamageTaken' 1`] = `"You take #% reduced Extra Damage from Critical Strikes"`;

exports[`usage with poe-mods matches the snapshot of 'ReflectedDamage' 1`] = `"#% reduced Reflected * Damage taken"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedChaosDamage' 1`] = `"Adds # to # Chaos Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedChaosDamageHybrid' 1`] = `"Adds # to # Chaos Damage to Spells / #% increased Chaos Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedColdDamageHybrid' 1`] = `"Adds # to # Cold Damage to Spells / #% increased Cold Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedColdDamageUber' 1`] = `"Adds # to # Cold Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedElementalDamage' 1`] = `"Adds # to # * Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedFireDamageHybrid' 1`] = `"Adds # to # Fire Damage to Spells / #% increased Fire Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedFireDamageUber' 1`] = `"Adds # to # Fire Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedLightningDamageHybrid' 1`] = `"Adds # to # Lightning Damage to Spells / #% increased Lightning Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedLightningDamageUber' 1`] = `"Adds # to # Lightning Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedPhysicalDamage' 1`] = `"Adds # to # Physical Damage to Spells"`;

exports[`usage with poe-mods matches the snapshot of 'SpellAddedPhysicalDamageHybrid' 1`] = `"Adds # to # Physical Damage to Spells / #% increased Global Physical Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellDamage' 1`] = `"#% increased Spell Damage"`;

exports[`usage with poe-mods matches the snapshot of 'SpellDamageAndMana' 1`] = `"#% increased Spell Damage / +# to maximum Mana"`;

exports[`usage with poe-mods matches the snapshot of 'SpellDamagePerDexterity' 1`] = `"#% increased Spell Damage per 16 Dexterity"`;

exports[`usage with poe-mods matches the snapshot of 'SpellDamagePerIntelligence' 1`] = `"#% increased Spell Damage per * Intelligence"`;

exports[`usage with poe-mods matches the snapshot of 'SpellDamagePerStrength' 1`] = `"#% increased Spell Damage per * Strength"`;

exports[`usage with poe-mods matches the snapshot of 'SupportGemQuality' 1`] = `"+#% to Quality of Socketed Support Gems"`;

exports[`usage with poe-mods matches the snapshot of 'Supported' 1`] = `"Socketed Gems are Supported by Level # *"`;

exports[`usage with poe-mods matches the snapshot of 'SupportedByLifeLeech' 1`] = `"Socketed Gems are supported by Level # Life Leech"`;

exports[`usage with poe-mods matches the snapshot of 'SupportedByTrigger' 1`] = `"Socketed Gems are * by Level # Cast * * *"`;

exports[`usage with poe-mods matches the snapshot of 'WeaponCasterDamagePrefix' 1`] = `"*"`;
